### PR TITLE
RavenDB-12917 Merge database record configuration in import

### PR DESCRIPTION
**Pull request by efratshenhar**

### CODE REVIEW

File: src/Raven.Server/ServerWide/ClusterStateMachine.cs [hunk 1]
```diff
@@ -229,7 +229,8 @@ protected override void Apply(TransactionOperationContext context, BlittableJson
                     case nameof(RemoveSqlConnectionStringCommand):
                     case nameof(UpdatePullReplicationAsHubCommand):
                     case nameof(UpdatePullReplicationAsSinkCommand):
-                        UpdateDatabase(context, type, cmd, index, leader, serverStore);
+                    case nameof(EditDatabaseClientConfigurationCommand):
+                    UpdateDatabase(context, type, cmd, index, leader, serverStore);
```
Comment (ayende) [inline src/Raven.Server/ServerWide/ClusterStateMachine.cs]:
```suggestion
                        UpdateDatabase(context, type, cmd, index, leader, serverStore);
```

File: src/Raven.Server/Smuggler/Documents/DatabaseDestination.cs [hunk 2]
```diff
@@ -533,9 +534,19 @@ public void WriteDatabaseRecord(DatabaseRecord databaseRecord, SmugglerProgressB
...
 
-                if (currentDatabaseRecord?.ConflictSolverConfig == null &&
-                    databaseRecord?.ConflictSolverConfig != null)
+                if (databaseRecord?.ConflictSolverConfig != null)
                 {
+                    if (currentDatabaseRecord?.ConflictSolverConfig != null)
+                    {
+                        foreach (var collection in currentDatabaseRecord.ConflictSolverConfig.ResolveByCollection)
+                        {
+                            if (!(databaseRecord.ConflictSolverConfig.ResolveByCollection.ContainsKey(collection.Key)))
```
Comment (ayende) [inline src/Raven.Server/Smuggler/Documents/DatabaseDestination.cs]: Shouldn't we always overwrite it here?

File: src/Raven.Server/Smuggler/Documents/DatabaseDestination.cs [hunk 3]
```diff
@@ -545,26 +556,41 @@ public void WriteDatabaseRecord(DatabaseRecord databaseRecord, SmugglerProgressB
...
 
-                if (currentDatabaseRecord?.PeriodicBackups.Count == 0 &&
-                    databaseRecord?.PeriodicBackups.Count > 0)
+                if (databaseRecord?.PeriodicBackups.Count > 0)
                 {
                     if (_log.IsInfoEnabled)
                         _log.Info("Configuring periodic backups configuration from smuggler");
                     foreach (var backupConfig in databaseRecord.PeriodicBackups)
                     {
+                        currentDatabaseRecord?.PeriodicBackups.ForEach(x =>
```
Comment (ayende) [inline src/Raven.Server/Smuggler/Documents/DatabaseDestination.cs:565]: If the `currentDatabaseRecord` is null, this won't get executed at all, which I assume isn't intended.
In that case, won't we want to add _all_ of them?

File: src/Raven.Server/Smuggler/Documents/DatabaseDestination.cs [hunk 4]
```diff
@@ -545,26 +556,41 @@ public void WriteDatabaseRecord(DatabaseRecord databaseRecord, SmugglerProgressB
...
 
-                if (currentDatabaseRecord?.SinkPullReplications.Count == 0 &&
-                    databaseRecord?.SinkPullReplications.Count > 0)
+                if (databaseRecord?.SinkPullReplications.Count > 0)
                 {
                     if (_log.IsInfoEnabled)
                         _log.Info("Configuring sink pull replication configuration from smuggler");
                     foreach (var pullReplication in databaseRecord.SinkPullReplications)
                     {
+                        currentDatabaseRecord?.SinkPullReplications.ForEach(x =>
```
Comment (ayende) [inline src/Raven.Server/Smuggler/Documents/DatabaseDestination.cs:586]: Same

File: src/Raven.Server/Smuggler/Documents/DatabaseDestination.cs [hunk 5]
```diff
@@ -574,13 +600,20 @@ public void WriteDatabaseRecord(DatabaseRecord databaseRecord, SmugglerProgressB
...
 
-                if (currentDatabaseRecord?.HubPullReplications.Count == 0 &&
-                    databaseRecord?.HubPullReplications.Count > 0)
+                if (databaseRecord?.HubPullReplications.Count > 0)
                 {
                     if (_log.IsInfoEnabled)
                         _log.Info("Configuring hub pull replication configuration from smuggler");
                     foreach (var pullReplication in databaseRecord.HubPullReplications)
                     {
+                        currentDatabaseRecord?.HubPullReplications.ForEach(x =>
```
Comment (ayende) [inline src/Raven.Server/Smuggler/Documents/DatabaseDestination.cs:609]: same

File: src/Raven.Server/Smuggler/Documents/DatabaseDestination.cs [hunk 6]
```diff
@@ -605,13 +637,20 @@ public void WriteDatabaseRecord(DatabaseRecord databaseRecord, SmugglerProgressB
...
 
-                if (currentDatabaseRecord?.ExternalReplications.Count == 0 &&
-                    databaseRecord?.ExternalReplications.Count > 0)
+                if (databaseRecord?.ExternalReplications.Count > 0)
                 {
                     if (_log.IsInfoEnabled)
                         _log.Info("Configuring external replications configuration from smuggler");
                     foreach (var replication in databaseRecord.ExternalReplications)
                     {
+                        currentDatabaseRecord?.ExternalReplications.ForEach(x =>
```
Comment (ayende) [inline src/Raven.Server/Smuggler/Documents/DatabaseDestination.cs:646]: Same

File: src/Raven.Server/Smuggler/Documents/DatabaseDestination.cs [hunk 7]
```diff
@@ -621,52 +660,73 @@ public void WriteDatabaseRecord(DatabaseRecord databaseRecord, SmugglerProgressB
...
 
-                if (currentDatabaseRecord?.RavenEtls.Count == 0 &&
-                    databaseRecord?.RavenEtls.Count > 0)
+                if (databaseRecord?.RavenEtls.Count > 0)
                 {
                     if (_log.IsInfoEnabled)
                         _log.Info("Configuring raven etls configuration from smuggler");
                     foreach (var etl in databaseRecord.RavenEtls)
                     {
+                        currentDatabaseRecord?.RavenEtls.ForEach(x =>
```
Comment (ayende) [inline src/Raven.Server/Smuggler/Documents/DatabaseDestination.cs:669]: Same

File: src/Raven.Server/Smuggler/Documents/DatabaseDestination.cs [hunk 8]
```diff
@@ -621,52 +660,73 @@ public void WriteDatabaseRecord(DatabaseRecord databaseRecord, SmugglerProgressB
...
 
-                if (currentDatabaseRecord?.SqlEtls.Count == 0 &&
-                    databaseRecord?.SqlEtls.Count > 0)
+                if (databaseRecord?.SqlEtls.Count > 0)
                 {
                     if (_log.IsInfoEnabled)
                         _log.Info("Configuring sql etls configuration from smuggler");
                     foreach (var etl in databaseRecord.SqlEtls)
                     {
+                        currentDatabaseRecord?.SqlEtls.ForEach(x =>
```
Comment (ayende) [inline src/Raven.Server/Smuggler/Documents/DatabaseDestination.cs:689]: Same

File: src/Raven.Server/Smuggler/Documents/DatabaseDestination.cs [hunk 9]
```diff
@@ -621,52 +660,73 @@ public void WriteDatabaseRecord(DatabaseRecord databaseRecord, SmugglerProgressB
...
 
-                if (currentDatabaseRecord?.Revisions == null &&
-                    databaseRecord?.Revisions != null)
+                if (databaseRecord?.Revisions != null)
                 {
+                    if (currentDatabaseRecord?.Revisions != null)
+                    {
+                        foreach (var collection in currentDatabaseRecord.Revisions.Collections)
+                        {
+                            if (!(databaseRecord.Revisions.Collections.ContainsKey(collection.Key)))
```
Comment (ayende) [inline src/Raven.Server/Smuggler/Documents/DatabaseDestination.cs]: Don't we want to overwrite this?

**Review by ayende — CHANGES_REQUESTED** [review]

File: src/Raven.Server/ServerWide/Commands/EditDatabaseClientConfigurationCommand.cs [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+using Raven.Client.Documents.Operations.Revisions;
+using Raven.Client.ServerWide;
+using Raven.Server.Utils;
+using Sparrow.Json.Parsing;
+
+namespace Raven.Server.ServerWide.Commands
+{
+    public class EditDatabaseClientConfigurationCommand : UpdateDatabaseCommand
+    {
+        public ClientConfiguration Configuration { get; protected set; }
```
Comment (ppekrol) [inline src/Raven.Server/ServerWide/Commands/EditDatabaseClientConfigurationCommand.cs]: I think that this needs to be public `set` for proper deserialization. Or a field.

File: src/Raven.Server/Smuggler/Documents/DatabaseDestination.cs [hunk 11]
```diff
@@ -545,26 +556,41 @@ public void WriteDatabaseRecord(DatabaseRecord databaseRecord, SmugglerProgressB
...
-                    databaseRecord?.PeriodicBackups.Count > 0)
+                if (databaseRecord?.PeriodicBackups.Count > 0)
                 {
                     if (_log.IsInfoEnabled)
                         _log.Info("Configuring periodic backups configuration from smuggler");
                     foreach (var backupConfig in databaseRecord.PeriodicBackups)
                     {
+                        currentDatabaseRecord?.PeriodicBackups.ForEach(x =>
+                        {
+                            if (x.Name.Equals(backupConfig.Name, StringComparison.InvariantCultureIgnoreCase))
```
Comment (ppekrol) [inline src/Raven.Server/Smuggler/Documents/DatabaseDestination.cs]: `StringComparison.OridinalIgnoreCase`

File: src/Raven.Server/Smuggler/Documents/DatabaseDestination.cs [hunk 12]
```diff
@@ -545,26 +556,41 @@ public void WriteDatabaseRecord(DatabaseRecord databaseRecord, SmugglerProgressB
...
-                    databaseRecord?.SinkPullReplications.Count > 0)
+                if (databaseRecord?.SinkPullReplications.Count > 0)
                 {
                     if (_log.IsInfoEnabled)
                         _log.Info("Configuring sink pull replication configuration from smuggler");
                     foreach (var pullReplication in databaseRecord.SinkPullReplications)
                     {
+                        currentDatabaseRecord?.SinkPullReplications.ForEach(x =>
+                        {
+                            if (x.Name.Equals(pullReplication.Name, StringComparison.InvariantCultureIgnoreCase))
```
Comment (ppekrol) [inline src/Raven.Server/Smuggler/Documents/DatabaseDestination.cs]: `StringComparison.OridinalIgnoreCase`

**Review by ppekrol — COMMENTED** [review]
@efratshenhar ping me on Sunday about this

File: src/Raven.Server/ServerWide/Commands/EditDatabaseClientConfigurationCommand.cs [hunk 13]
```diff
@@ -0,0 +1,38 @@
...
+using Raven.Server.Utils;
+using Sparrow.Json.Parsing;
+
+namespace Raven.Server.ServerWide.Commands
+{
+    public class EditDatabaseClientConfigurationCommand : UpdateDatabaseCommand
+    {
+        public ClientConfiguration Configuration { get; set; }
+
+        public void UpdateDatabaseRecord(DatabaseRecord databaseRecord)
```
Comment (ppekrol) [inline src/Raven.Server/ServerWide/Commands/EditDatabaseClientConfigurationCommand.cs]: Is this method used?

File: src/Raven.Server/Smuggler/Documents/DatabaseDestination.cs [hunk 14]
```diff
@@ -533,9 +534,19 @@ public void WriteDatabaseRecord(DatabaseRecord databaseRecord, SmugglerProgressB
...
 
-                if (currentDatabaseRecord?.ConflictSolverConfig == null &&
-                    databaseRecord?.ConflictSolverConfig != null)
+                if (databaseRecord?.ConflictSolverConfig != null)
                 {
+                    if (currentDatabaseRecord?.ConflictSolverConfig != null)
+                    {
+                        foreach (var collection in currentDatabaseRecord.ConflictSolverConfig.ResolveByCollection)
+                        {
+                            if ((databaseRecord.ConflictSolverConfig.ResolveByCollection.ContainsKey(collection.Key)) == false)
```
Comment (ppekrol) [inline src/Raven.Server/Smuggler/Documents/DatabaseDestination.cs:543]: Do we need to do OridinalIgnoreCase here?

File: src/Raven.Server/Smuggler/Documents/DatabaseDestination.cs [hunk 15]
```diff
@@ -574,13 +600,20 @@ public void WriteDatabaseRecord(DatabaseRecord databaseRecord, SmugglerProgressB
...
-                    databaseRecord?.HubPullReplications.Count > 0)
+                if (databaseRecord?.HubPullReplications.Count > 0)
                 {
                     if (_log.IsInfoEnabled)
                         _log.Info("Configuring hub pull replication configuration from smuggler");
                     foreach (var pullReplication in databaseRecord.HubPullReplications)
                     {
+                        currentDatabaseRecord?.HubPullReplications.ForEach(x =>
+                        {
+                            if (x.Name.Equals(pullReplication.Name, StringComparison.InvariantCultureIgnoreCase))
```
Comment (ppekrol) [inline src/Raven.Server/Smuggler/Documents/DatabaseDestination.cs]: OridinalIgnoreCase

File: src/Raven.Server/Smuggler/Documents/DatabaseDestination.cs [hunk 16]
```diff
@@ -605,13 +637,20 @@ public void WriteDatabaseRecord(DatabaseRecord databaseRecord, SmugglerProgressB
...
-                    databaseRecord?.ExternalReplications.Count > 0)
+                if (databaseRecord?.ExternalReplications.Count > 0)
                 {
                     if (_log.IsInfoEnabled)
                         _log.Info("Configuring external replications configuration from smuggler");
                     foreach (var replication in databaseRecord.ExternalReplications)
                     {
+                        currentDatabaseRecord?.ExternalReplications.ForEach(x =>
+                        {
+                            if (x.Name.Equals(replication.Name))
```
Comment (ppekrol) [inline src/Raven.Server/Smuggler/Documents/DatabaseDestination.cs]: Do we need to do OridinalIgnoreCase here?

File: src/Raven.Server/Smuggler/Documents/DatabaseDestination.cs [hunk 17]
```diff
@@ -621,52 +660,73 @@ public void WriteDatabaseRecord(DatabaseRecord databaseRecord, SmugglerProgressB
...
-                    databaseRecord?.RavenEtls.Count > 0)
+                if (databaseRecord?.RavenEtls.Count > 0)
                 {
                     if (_log.IsInfoEnabled)
                         _log.Info("Configuring raven etls configuration from smuggler");
                     foreach (var etl in databaseRecord.RavenEtls)
                     {
+                        currentDatabaseRecord?.RavenEtls.ForEach(x =>
+                        {
+                            if (x.Name.Equals(etl.Name, StringComparison.InvariantCultureIgnoreCase))
```
Comment (ppekrol) [inline src/Raven.Server/Smuggler/Documents/DatabaseDestination.cs]: OridinalIgnoreCase

File: src/Raven.Server/Smuggler/Documents/DatabaseDestination.cs [hunk 18]
```diff
@@ -621,52 +660,73 @@ public void WriteDatabaseRecord(DatabaseRecord databaseRecord, SmugglerProgressB
...
-                    databaseRecord?.SqlEtls.Count > 0)
+                if (databaseRecord?.SqlEtls.Count > 0)
                 {
                     if (_log.IsInfoEnabled)
                         _log.Info("Configuring sql etls configuration from smuggler");
                     foreach (var etl in databaseRecord.SqlEtls)
                     {
+                        currentDatabaseRecord?.SqlEtls.ForEach(x =>
+                        {
+                            if (x.Name.Equals(etl.Name, StringComparison.InvariantCultureIgnoreCase))
```
Comment (ppekrol) [inline src/Raven.Server/Smuggler/Documents/DatabaseDestination.cs]: OridinalIgnoreCase

File: src/Raven.Server/Smuggler/Documents/DatabaseDestination.cs [hunk 19]
```diff
@@ -621,52 +660,73 @@ public void WriteDatabaseRecord(DatabaseRecord databaseRecord, SmugglerProgressB
...
 
-                if (currentDatabaseRecord?.Revisions == null &&
-                    databaseRecord?.Revisions != null)
+                if (databaseRecord?.Revisions != null)
                 {
+                    if (currentDatabaseRecord?.Revisions != null)
+                    {
+                        foreach (var collection in currentDatabaseRecord.Revisions.Collections)
+                        {
+                            if ((databaseRecord.Revisions.Collections.ContainsKey(collection.Key)) == false)
```
Comment (ppekrol) [inline src/Raven.Server/Smuggler/Documents/DatabaseDestination.cs:709]: Do we need to do OridinalIgnoreCase here?

**Review by ppekrol — CHANGES_REQUESTED** [review]

File: src/Raven.Server/Smuggler/Documents/DatabaseDestination.cs (hunk 19 above)
Comment (efratshenhar) [reply src/Raven.Server/Smuggler/Documents/DatabaseDestination.cs:709]: No

File: src/Raven.Server/Smuggler/Documents/DatabaseDestination.cs (hunk 14 above)
Comment (efratshenhar) [reply src/Raven.Server/Smuggler/Documents/DatabaseDestination.cs:543]: No

File: src/Raven.Server/ServerWide/Commands/EditDatabaseClientConfigurationCommand.cs (hunk 13 above)
Comment (efratshenhar) [reply src/Raven.Server/ServerWide/Commands/EditDatabaseClientConfigurationCommand.cs]: No

File: src/Raven.Server/Smuggler/Documents/DatabaseDestination.cs (hunk 16 above)
Comment (efratshenhar) [reply src/Raven.Server/Smuggler/Documents/DatabaseDestination.cs]: Yes

**Review by ppekrol — APPROVED** [review]

**Pull request merged**